fabric-rendering-v1/src/client/java/net/fabricmc/fabric/impl/client/rendering/LayeredDrawerWrapperImpl.java

```diff
@@ -0,0 +1,167 @@
+			iterator.add(layer);
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + afterThis + " not found");
+		}
+
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper addLayerBefore(Identifier beforeThis, IdentifiedLayer layer) {
+		validateUnique(layer);
+		boolean didChange = findLayer(beforeThis, (l, iterator) -> {
+			iterator.previous();
+			iterator.add(layer);
+			iterator.next();
+			return true;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + beforeThis + " not found");
+		}
+
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper removeLayer(Identifier identifier) {
+		boolean didChange = findLayer(identifier, (l, iterator) -> {
+			iterator.remove();
+			return false;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + identifier + " not found");
+		}
+
+		return this;
+	}
+
+	@Override
+	public LayeredDrawerWrapper replaceLayer(Identifier identifier, Function<IdentifiedLayer, IdentifiedLayer> replacer) {
+		boolean didChange = findLayer(identifier, (l, iterator) -> {
+			iterator.remove();
+			iterator.add(replacer.apply((IdentifiedLayer) l));
+			return false;
+		});
+
+		if (!didChange) {
+			throw new IllegalArgumentException("Layer with identifier " + identifier + " not found");
+		}
+
+		return this;
+	}
+
+	private void validateUnique(IdentifiedLayer layer) {
+		visitLayers((l, iterator) -> {
+			if (matchesIdentifier(l, layer.id())) {
+				throw new IllegalArgumentException("Layer with identifier " + layer.id() + " already exists");
+			}
+
+			return true;
+		});
+	}
+
+	private boolean findLayer(Identifier identifier, LayerVisitor visitor) {
+		AtomicBoolean didFind = new AtomicBoolean(false);
+
+		visitLayers((l, iterator) -> {
+			if (matchesIdentifier(l, identifier)) {
+				didFind.set(true);
+				visitor.visit(l, iterator);
+			}
+
+			return true;
+		});
+
+		return didFind.get();
+	}
+
+	private void visitLayers(LayerVisitor visitor) {
+		visitLayers(getLayers(base), visitor);
+	}
+
+	private void visitLayers(List<LayeredDrawer.Layer> layers, LayerVisitor visitor) {
+		ListIterator<LayeredDrawer.Layer> iterator = layers.listIterator();
+
+		while (iterator.hasNext()) {
+			LayeredDrawer.Layer layer = iterator.next();
+
+			if (!visitor.visit(layer, iterator)) {
+				iterator.remove();
+				continue;
+			}
+
+			if (layer instanceof SubLayer subLayer) {
+				visitLayers(getLayers(subLayer.delegate()), visitor);
+			}
+		}
+	}
+
+	private static boolean matchesIdentifier(LayeredDrawer.Layer layer, Identifier identifier) {
+		return layer instanceof IdentifiedLayer il && il.id().equals(identifier);
+	}
+
+	private interface LayerVisitor {
+		/**
+		 * @return false to remove the layer, true to keep it
+		 */
+		boolean visit(LayeredDrawer.Layer layer, ListIterator<LayeredDrawer.Layer> iterator);
+	}
```

Comment:
I cannot grasp the meaning of `LayerVisitor#visit` having a return value.

Look at removeLayer for example. When a layer is found, `iterator.remove()` is called. Makes sense, it should be removed from the list.

But then `false` is returned, which will result in another call to `iterator.remove()` by `visitLayers`, why? Either the `LayerVisitor` removes the item or `visitLayers` does. Doing this twice seems very confusing to me

Reply:
Yes, something isnt right here. I think the return can be removed I think it was from an eariler example of this code. The unit tests clearly need expanding to test this.

Reply:
Huh, how did the unit test succeed just fine?

Reply:
Oh I get it. It currently works, just a bit lacking in documentation. `removeLayer` calls `findLayer` with a lambda that returns false, but `findLayer` always calls `visitLayers` with a lambda that returns true and ignore the return value of the passed in lambda.